File: sendfn/typescript/src/push/service.ts
import {
  SendPushParams,
  PushNotification,
  SendfnOptions,
  Platform
} from '../types';
import { PushProviderError } from '../errors';
import { SendfnDatabaseAdapter } from '../database/adapter';
import { PushProvider } from './provider';
import { DeviceTokenManager } from './device-manager';
import { mapWithConcurrency, resolveConcurrency } from '../utils/concurrency';

const PLATFORM_ORDER: Platform[] = ['android', 'ios', 'web'];
const DEFAULT_BULK_CONCURRENCY = 5;

export class PushService {
  constructor(
    private providers: Map<Platform, PushProvider>,
    private adapter: SendfnDatabaseAdapter,
    private deviceManager: DeviceTokenManager,
    private options: SendfnOptions
  ) {}

  async sendPush(params: SendPushParams): Promise<PushNotification> {
    // 1. Resolve tokens
    const platformTokenSets = new Map<Platform, Set<string>>();

    // If userId provided, fetch tokens
    const userIds = Array.isArray(params.userId) ? params.userId : [params.userId];
    const recipientMetadata = { ...(params.metadata || {}), recipientUserIds: [...userIds] };

    // This is simplified. Ideally we fetch all tokens for all users.
    for (const uid of userIds) {
        const devices = await this.deviceManager.getActiveDevices(uid);
        for (const device of devices) {
            const tokens = platformTokenSets.get(device.platform) ?? new Set<string>();
            tokens.add(device.token);
            platformTokenSets.set(device.platform, tokens);
        }
    }

    if (platformTokenSets.size === 0) {
        // Create a failed notification record or just return?
        // Let's create a failed one to track intent.
        const notification = await this.adapter.createPushNotification({
            userId: userIds.join(','),
            title: params.title,
            body: params.body,
            data: params.data || null,
            deviceTokens: [],
            platform: 'web', // Default/Unknown
            provider: 'none',
            status: 'failed',
            sentCount: 0,
            failedCount: 0,
            sentAt: null,
            metadata: { ...recipientMetadata, error: 'No active devices found' }
        });
        return notification;
    }

    // 2. Create PushNotification record(s)
    // The schema assumes single provider per record?
    // "provider: string (e.g., "fcm", "apns")"
    // If we send to mixed platforms, we might need multiple records or a "mixed" provider.
    // For v0, let's create one record per platform send to keep it clean.
    // But the method returns Promise<PushNotification>.
    // Let's assume we return the "primary" one or aggregate.
    // Actually, if we have multiple platforms, we should probably treat them as separate operations internally
    // but the API call is single.
    // Let's handle just one for simplicity in return type, or change return type.
    // The spec says: "Return PushNotification".

    // Let's iterate platforms and send.

    const platforms = PLATFORM_ORDER.filter((platform) => (platformTokenSets.get(platform)?.size ?? 0) > 0);
    const missingPlatform = platforms.find((platform) => !this.providers.has(platform));
    if (missingPlatform) {
      throw new PushProviderError(`No push provider configured for platform ${missingPlatform}`, {
        retryable: false,
      });
    }
    let logicalNotificationId: string | null = null;
    let firstNotificationId: string | null = null;
    let aggregateSentCount = 0;
    let aggregateFailedCount = 0;
    let logicalMetadata: Record<string, any> = {
      ...recipientMetadata,
      notificationIds: [] as string[],
    };
    let logicalSentAt: Date | null = null;
    let firstProviderError: unknown;
    const providerErrors: Array<{ platform: Platform; provider: string; error: string }> = [];

    for (const platform of platforms) {
        const pTokens = [...platformTokenSets.get(platform)!];
        const provider = this.providers.get(platform)!;

        const notification = await this.adapter.createPushNotification({
            userId: userIds.join(','),
            title: params.title,
            body: params.body,
            data: params.data || null,
            deviceTokens: pTokens,
            platform: platform,
            provider: provider.name,
            status: 'pending',
            sentCount: 0,
            failedCount: 0,
            sentAt: null,
            metadata: recipientMetadata
        });

        (logicalMetadata.notificationIds as string[]).push(notification.id);
        firstNotificationId ??= notification.id;

        let response: Awaited<ReturnType<PushProvider['sendPush']>>;
        try {
            response = await provider.sendPush({
                platform,
                deviceTokens: pTokens,
                title: params.title,
                body: params.body,
                data: params.data,
                imageUrl: params.imageUrl,
                badge: params.badge,
                sound: params.sound,
                priority: params.priority,
                ttl: params.ttl,
                collapseKey: params.collapseKey,
                category: params.category
            });

        } catch (error: any) {
             await this.adapter.updatePushNotification(notification.id, {
                status: 'failed',
                metadata: { ...recipientMetadata, error: error.message }
            });

             if (this.options.eventTracking !== false) await this.adapter.recordEvent({
                referenceId: notification.id,
                referenceType: 'push',
                eventType: 'failed',
                provider: provider.name,
                providerEventId: null,
                recipientEmail: null,
                recipientPhone: null,
                deviceToken: null,
                metadata: { error: error.message },
                eventTimestamp: new Date()
            });

            aggregateFailedCount += pTokens.length;
            firstProviderError ??= error;
            providerErrors.push({ platform, provider: provider.name, error: error.message });
            continue;
        }

        if (response.invalidTokens.length > 0) {
            await this.deviceManager.deactivateTokens(response.invalidTokens);
        }

        const delivered = response.successCount > 0;
        await this.adapter.updatePushNotification(notification.id, {
            status: delivered ? 'sent' : 'failed',
            sentCount: response.successCount,
            failedCount: response.failedCount,
            sentAt: response.timestamp,
            metadata: {
                ...recipientMetadata,
                results: response.results
            }
        });

        if (this.options.eventTracking !== false) await this.adapter.recordEvent({
            referenceId: notification.id,
            referenceType: 'push',
            eventType: delivered ? 'sent' : 'failed',
            provider: provider.name,
            providerEventId: null,
            recipientEmail: null,
            recipientPhone: null,
            deviceToken: null,
            metadata: {
                successCount: response.successCount,
                failedCount: response.failedCount
            },
            eventTimestamp: response.timestamp
        });

        aggregateSentCount += response.successCount;
        aggregateFailedCount += response.failedCount;
        if (delivered) logicalNotificationId ??= notification.id;
        if (!logicalSentAt) {
          logicalSentAt = response.timestamp;
        }
    }

    logicalNotificationId ??= firstNotificationId;
    if (!logicalNotificationId) {
        throw new PushProviderError("Failed to process push for any platform");
    }

    const existingLogicalNotification = await this.adapter.getPushNotification(logicalNotificationId);
    const logicalNotification = await this.adapter.updatePushNotification(logicalNotificationId, {
      status: aggregateSentCount > 0 ? 'sent' : 'failed',
      sentCount: aggregateSentCount,
      failedCount: aggregateFailedCount,
      sentAt: logicalSentAt,
      metadata: {
        ...(existingLogicalNotification?.metadata || {}),
        ...logicalMetadata,
        ...(providerErrors.length > 0 ? { providerErrors } : {}),
      },
    });

    if (aggregateSentCount === 0 && firstProviderError) {
      throw firstProviderError;
    }
    return logicalNotification;
  }

  async sendBulkPush(notifications: SendPushParams[]): Promise<PushNotification[]> {
      return mapWithConcurrency(
        notifications,
        resolveConcurrency(this.options.bulkConcurrency, DEFAULT_BULK_CONCURRENCY),
        async (notification) => this.sendPush(notification)
      );
  }
}

File: sendfn/python/sendfn/push/service.py
"""Push notification service orchestration."""

from datetime import datetime
from typing import Optional

from superfunctions.db import Adapter

from .._concurrency import map_with_concurrency, resolve_concurrency
from ..errors import PushProviderError
from ..models import Platform, PushNotification, SendPushParams
from .device_manager import DeviceTokenManager
from .provider import PushProvider

PLATFORM_ORDER: tuple[Platform, ...] = ("android", "ios", "web")


class PushService:
    """Push notification service that coordinates providers and device management."""

    def __init__(
        self,
        providers: dict[Platform, PushProvider],
        db: Adapter,
        device_manager: DeviceTokenManager,
        bulk_concurrency: int = 5,
        event_tracking: bool = True,
    ) -> None:
        """Initialize push service.

        Args:
            providers: Map of platform to push providers (FCM, APNS)
            db: Database adapter
            device_manager: Device token manager
        """
        self.providers = providers
        self.db = db
        self.device_manager = device_manager
        self.bulk_concurrency = resolve_concurrency(bulk_concurrency, 5)
        self.event_tracking = event_tracking

    async def send_push(self, params: SendPushParams) -> PushNotification:
        """Send a push notification.

        Args:
            params: Push send parameters

        Returns:
            Push notification record
        """
        from ..database.helpers import (
            create_push_notification,
            get_push_notification,
            record_event,
            update_push_notification,
        )

        # Normalize user IDs
        user_ids = list(
            dict.fromkeys(params.user_id if isinstance(params.user_id, list) else [params.user_id])
        )

        # Resolve user IDs to device tokens
        tokens: list[str] = []
        platform_tokens: dict[Platform, list[str]] = {}
        platform_token_sets: dict[Platform, set[str]] = {}
        platform_recipient_sets: dict[Platform, set[str]] = {}

        for user_id in user_ids:
            devices = await self.device_manager.get_active_devices(user_id)
            for device in devices:
                platform_recipient_sets.setdefault(device.platform, set()).add(user_id)
                seen_tokens = platform_token_sets.setdefault(device.platform, set())
                if device.token in seen_tokens:
                    continue
                seen_tokens.add(device.token)
                tokens.append(device.token)
                if device.platform not in platform_tokens:
                    platform_tokens[device.platform] = []
                platform_tokens[device.platform].append(device.token)

        # If no devices found, create failed notification
        if not tokens:
            notification = await create_push_notification(
                self.db,
                {
                    "userId": ",".join(user_ids),
                    "title": params.title,
                    "body": params.body,
                    "data": params.data,
                    "deviceTokens": [],
                    "platform": "web",  # Default
                    "provider": "none",
                    "status": "failed",
                    "sentCount": 0,
                    "failedCount": 0,
                    "sentAt": None,
                    "metadata": {
                        **(params.metadata or {}),
                        "recipientUserIds": user_ids,
                        "error": "No active devices found",
                    },
                },
            )
            return notification

        ordered_platforms = [
            platform for platform in PLATFORM_ORDER if platform_tokens.get(platform)
        ]
        missing_platform = next(
            (platform for platform in ordered_platforms if platform not in self.providers),
            None,
        )
        if missing_platform is not None:
            raise PushProviderError(
                f"No push provider configured for platform {missing_platform}",
                retryable=False,
            )
        logical_notification_id: Optional[str] = None
        first_notification_id: Optional[str] = None
        notification_ids: list[str] = []
        aggregate_sent_count = 0
        aggregate_failed_count = 0
        logical_sent_at: Optional[datetime] = None
        first_provider_error: Optional[Exception] = None
        provider_errors: list[dict[str, str]] = []
        bookkeeping_errors: list[dict[str, str]] = []
        notification_recipient_ids: dict[str, list[str]] = {}

        def record_bookkeeping_error(
            platform: Platform,
            provider: str,
            stage: str,
            error: Exception,
        ) -> None:
            bookkeeping_errors.append(
                {
                    "platform": platform,
                    "provider": provider,
                    "stage": stage,
                    "error": str(error),
                }
            )

        for platform in ordered_platforms:
            p_tokens = platform_tokens[platform]
            platform_recipient_ids = [
                user_id
                for user_id in user_ids
                if user_id in platform_recipient_sets.get(platform, set())
            ]
            provider = self.providers.get(platform)

            assert provider is not None

            # Create notification record
            notification = await create_push_notification(
                self.db,
                {
                    "userId": ",".join(user_ids),
                    "title": params.title,
                    "body": params.body,
                    "data": params.data,
                    "deviceTokens": p_tokens,
                    "platform": platform,
                    "provider": provider.name,
                    "status": "pending",
                    "sentCount": 0,
                    "failedCount": 0,
                    "sentAt": None,
                    "metadata": {
                        **(params.metadata or {}),
                        "recipientUserIds": platform_recipient_ids,
                    },
                },
            )
            notification_ids.append(str(notification.id))
            notification_recipient_ids[str(notification.id)] = platform_recipient_ids
            if first_notification_id is None:
                first_notification_id = str(notification.id)

            from .provider import SendPushRequest

            try:
                response = await provider.send_push(
                    SendPushRequest(
                        device_tokens=p_tokens,
                        title=params.title,
                        body=params.body,
                        platform=platform,
                        data=params.data,
                        image_url=params.image_url,
                        badge=params.badge,
                        sound=params.sound,
                        priority=params.priority,
                        ttl=params.ttl,
                        collapse_key=params.collapse_key,
                        category=params.category,
                    )
                )

            except Exception as error:
                try:
                    await update_push_notification(
                        self.db,
                        str(notification.id),
                        {
                            "status": "failed",
                            "metadata": {
                                **(params.metadata or {}),
                                "recipientUserIds": platform_recipient_ids,
                                "error": str(error),
                            },
                        },
                    )
                except Exception as bookkeeping_error:
                    record_bookkeeping_error(
                        platform, provider.name, "notification:update-failed", bookkeeping_error
                    )

                if self.event_tracking:
                    try:
                        await record_event(
                            self.db,
                            {
                                "referenceId": str(notification.id),
                                "referenceType": "push",
                                "eventType": "failed",
                                "provider": provider.name,
                                "providerEventId": None,
                                "recipientEmail": None,
                                "recipientPhone": None,
                                "deviceToken": None,
                                "metadata": {
                                    "recipientUserIds": platform_recipient_ids,
                                    "error": str(error),
                                },
                                "eventTimestamp": datetime.now(),
                            },
                        )
                    except Exception as bookkeeping_error:
                        record_bookkeeping_error(
                            platform, provider.name, "event:failed", bookkeeping_error
                        )

                aggregate_failed_count += len(p_tokens)
                if first_provider_error is None:
                    first_provider_error = error
                provider_errors.append(
                    {"platform": platform, "provider": provider.name, "error": str(error)}
                )
                continue

            delivered = response.success_count > 0
            aggregate_sent_count += response.success_count
            aggregate_failed_count += response.failed_count
            if delivered and logical_notification_id is None:
                logical_notification_id = str(notification.id)
            if logical_sent_at is None:
                logical_sent_at = response.timestamp

            if response.invalid_tokens:
                try:
                    await self.device_manager.deactivate_tokens(response.invalid_tokens)
                except Exception as bookkeeping_error:
                    record_bookkeeping_error(
                        platform, provider.name, "tokens:deactivate", bookkeeping_error
                    )

            try:
                await update_push_notification(
                    self.db,
                    str(notification.id),
                    {
                        "status": "sent" if delivered else "failed",
                        "sentCount": response.success_count,
                        "failedCount": response.failed_count,
                        "sentAt": response.timestamp,
                        "metadata": {
                            **(params.metadata or {}),
                            "recipientUserIds": platform_recipient_ids,
                            "results": response.results,
                        },
                    },
                )
            except Exception as bookkeeping_error:
                record_bookkeeping_error(
                    platform, provider.name, "notification:update-sent", bookkeeping_error
                )

            if self.event_tracking:
                try:
                    await record_event(
                        self.db,
                        {
                            "referenceId": str(notification.id),
                            "referenceType": "push",
                            "eventType": "sent" if delivered else "failed",
                            "provider": provider.name,
                            "providerEventId": None,
                            "recipientEmail": None,
                            "recipientPhone": None,
                            "deviceToken": None,  # Multiple tokens
                            "metadata": {
                                "recipientUserIds": platform_recipient_ids,
                                "successCount": response.success_count,
                                "failedCount": response.failed_count,
                            },
                            "eventTimestamp": response.timestamp,
                        },
                    )
                except Exception as bookkeeping_error:
                    record_bookkeeping_error(
                        platform, provider.name, "event:sent", bookkeeping_error
                    )

        logical_notification_id = logical_notification_id or first_notification_id
        if not logical_notification_id:
            raise PushProviderError("Failed to process push for any platform")

        existing_logical_notification = await get_push_notification(
            self.db,
            logical_notification_id,
        )
        logical_notification = await update_push_notification(
            self.db,
            logical_notification_id,
            {
                "status": "sent" if aggregate_sent_count > 0 else "failed",
                "sentCount": aggregate_sent_count,
                "failedCount": aggregate_failed_count,
                "sentAt": logical_sent_at,
                "metadata": {
                    **(
                        existing_logical_notification.metadata
                        if existing_logical_notification
                        else {}
                    ),
                    **(params.metadata or {}),
                    "recipientUserIds": notification_recipient_ids.get(logical_notification_id, []),
                    "notificationIds": notification_ids,
                    **({"providerErrors": provider_errors} if provider_errors else {}),
                    **({"bookkeepingErrors": bookkeeping_errors} if bookkeeping_errors else {}),
                },
            },
        )

        if aggregate_sent_count == 0 and first_provider_error is not None:
            raise first_provider_error
        return logical_notification

    async def send_bulk_push(self, notifications: list[SendPushParams]) -> list[PushNotification]:
        """Send multiple push notifications.

        Args:
            notifications: List of push send parameters

        Returns:
            List of push notification records
        """
        return await map_with_concurrency(
            notifications,
            self.bulk_concurrency,
            lambda notification, _index: self.send_push(notification),
        )
